### DISCUSSION TITLE
Pyosmium Filter Buffer Full

### DISCUSSION THREAD

**Discussion by derkgonschor** [osmcode/pyosmium#200] [General]
Hello everybody,

During my master project, we are working with OSM datasets and were  looking for a new osm-filter to filter out nodes, ways etc. which we need for our work later on.
Here, we have now decided to use the pyosmium filter, which works  great up to a certain point.
My current problem is that after a certain time regardless of the file size and the number of points, the error occurs that he buffers are full: 
_Traceback (most recent call last):
  File "C:/Users/derkg/OneDrive/Desktop/Ma_Sc_Projekt/osm-filter/osmfilter_test.py", line 130, in <module>
    print("Anzahl an Knotenpunkten: " + str(h.num_nodes))
  File "C:/Users/derkg/OneDrive/Desktop/Ma_Sc_Projekt/osm-filter/osmfilter_test.py", line 73, in way
    if self.run == 1 and r.tags.get('route') == 'power':
RuntimeError: Osmium buffer is full_
Please find the code attached.
[osmfilter_test.zip](https://github.com/osmcode/pyosmium/files/8570905/osmfilter_test.zip)
Many thanks in advance!

Best regards,
Derk

**Comment by lonvia** [comment]
The problem is with the lines with `self.writer.close()` in the PowerHandler. `close()` closes the output file. If you try writing again with `self.writer.add_node()` etc.  after closing, you get the error message you are seeing. It could be admittedly a bit more helpful. Remove the `close()` calls and add a final `writer.close()` in the main program after everything is done and you will be fine.

Another hint: don't write a single handler that you call three times with different `run` values. Write three handlers, a separate handler for each stage. In each handler only implement the callback functions for the OSM types you really need. That way pyosmium will know that it can skip over the sections of the file that you don't need and your program will run a lot faster.
